### PR TITLE
Add `stop generating` functionality

### PR DESCRIPTION
**Pull request by pablodanswer**
## Description
[Provide a brief description of the changes in this PR]

## How Has This Been Tested?
[Describe the tests you ran to verify your changes]

## Accepted Risk
[Any know risks or failure modes to point out to reviewers]

## Related Issue(s)
[If applicable, link to the issue(s) this PR addresses]

## Checklist:
- [ ] All of the automated tests pass
- [ ] All PR comments are addressed and marked resolved
- [ ] If there are migrations, they have been rebased to latest main
- [ ] If there are new dependencies, they are added to the requirements
- [ ] If there are new environment variables, they are added to all of the deployment methods
- [ ] If there are new APIs that don't require auth, they are added to PUBLIC_ENDPOINT_SPECS
- [ ] Docker images build and basic functionalities work
- [ ] Author has done a final read through of the PR right before merge

### CODE REVIEW

File: backend/danswer/db/chat.py [hunk 1]
```diff
@@ -406,33 +434,54 @@ def create_new_chat_message(
     error: str | None = None,
     reference_docs: list[DBSearchDoc] | None = None,
     alternate_assistant_id: int | None = None,
-    # Maps the citation number [n] to the DB SearchDoc
```
Comment (Weves) [inline backend/danswer/db/chat.py]: nit: why is this comment removed?

File: backend/danswer/llm/answering/answer.py [hunk 2]
```diff
@@ -153,6 +156,7 @@ def __init__(
 
         self._return_contexts = return_contexts
         self.skip_gen_ai_answer_generation = skip_gen_ai_answer_generation
+        self._is_generating = True
```
Comment (Weves) [inline backend/danswer/llm/answering/answer.py]: nit: prefer `is_cancelled` since that feels more clear as to what's happening. `is_generating` feels like a state of the internal rather than something that is triggered by the outside world

File: backend/danswer/llm/answering/answer.py [hunk 3]
```diff
@@ -537,3 +555,13 @@ def citations(self) -> list[CitationInfo]:
                 citations.append(packet)
 
         return citations
+
+    @property
+    def is_generating(self) -> bool:
```
Comment (Weves) [inline backend/danswer/llm/answering/answer.py]: same here (`is_generating` -> `is_cancelled`)

File: backend/danswer/server/query_and_chat/chat_backend.py [hunk 4]
```diff
@@ -271,21 +272,21 @@ def delete_chat_session_by_id(
...
 ) -> StreamingResponse:
-    """This endpoint is both used for all the following purposes:
-    - Sending a new message in the session
-    - Regenerating a message in the session (just send the same one again)
-    - Editing a message (similar to regenerating but sending a different message)
-    - Kicking off a seeded chat session (set `use_existing_user_message`)
-
-    To avoid extra overhead/latency, this assumes (and checks) that previous messages on the path
-    have already been set as latest"""
+    if request.method == "GET":
```
Comment (Weves) [inline backend/danswer/server/query_and_chat/chat_backend.py]: why is this needed now?

File: web/src/app/chat/lib.tsx [hunk 5]
```diff
@@ -106,70 +109,87 @@ export async function* sendMessage({
...
-    const errorJson = await sendMessageResponse.json();
-    const errorMsg = errorJson.message || errorJson.detail || "";
-    throw Error(`Failed to send message - ${errorMsg}`);
+
+  if (!response.ok) {
+    throw new Error(`HTTP error! status: ${response.status}`);
   }
 
-  yield* handleStream<PacketType>(sendMessageResponse);
+  const reader = response.body?.getReader();
```
Comment (Weves) [inline web/src/app/chat/lib.tsx]: could we have this as a utility like before? E.g. `handleSSEStream` or similar?

**Review by Weves — CHANGES_REQUESTED** [review]
Overall looks good 🎉 

I am running into this issue when stopping a message in a brand new convo: 

https://github.com/user-attachments/assets/07fd5096-30d6-4874-8407-4364e54ee41a

File: backend/danswer/llm/answering/answer.py (hunk 2 above)
Comment (pablonyx) [reply backend/danswer/llm/answering/answer.py]: Very fair - let's make the verbiage clear!

File: backend/danswer/server/query_and_chat/chat_backend.py [hunk 6]
```diff
@@ -272,20 +273,12 @@ def delete_chat_session_by_id(
 
 
 @router.post("/send-message")
-def handle_new_chat_message(
+async def handle_new_chat_message(
```
Comment (Weves) [inline backend/danswer/server/query_and_chat/chat_backend.py]: is it possible to do this in a non-async way? I think this could have some serious performance consequences (e.g. since none of the DB / OpenAI calls are async, this could block the main process preventing other requests from going through)

File: backend/danswer/server/query_and_chat/chat_backend.py [hunk 7]
```diff
@@ -272,20 +273,12 @@ def delete_chat_session_by_id(
...
     chat_message_req: CreateChatMessageRequest,
     request: Request,
     user: User | None = Depends(current_user),
     _: None = Depends(check_token_rate_limits),
 ) -> StreamingResponse:
-    """This endpoint is both used for all the following purposes:
```
Comment (Weves) [inline backend/danswer/server/query_and_chat/chat_backend.py:281]: nit: why is the comment removed 😅

File: web/src/app/chat/ChatPage.tsx [hunk 8]
```diff
@@ -442,7 +464,10 @@ export function ChatPage({
   const messageHistory = buildLatestMessageChain(
     completeMessageDetail.messageMap
   );
-  const [isStreaming, setIsStreaming] = useState(false);
+  const [submittedMessage, setSubmittedMessage] = useState("");
+  const [chatState, setChatState] = useState<ChatState>("input");
+  // const [isLoadingResponse, setIsLoadingResponse] = useState(false);
```
Comment (Weves) [inline web/src/app/chat/ChatPage.tsx]: nit: remove

File: web/src/app/chat/ChatPage.tsx [hunk 9]
```diff
@@ -887,7 +872,76 @@ export function ChatPage({
...
+              continue;
+            }
+
+            const messageResponseIDInfo = packet as MessageResponseIDInfo;
+
+            const user_message_id = messageResponseIDInfo.user_message_id!;
+            const assistant_message_id =
+              messageResponseIDInfo.reserved_assistant_message_id;
+
+            // if we're resending, set the parent's child to null
```
Comment (Weves) [inline web/src/app/chat/ChatPage.tsx]: nit: remove as this doesn't look relevant anymore?

**Review by Weves — CHANGES_REQUESTED** [review]
I am still running into the issue from https://github.com/danswer-ai/danswer/pull/2100#pullrequestreview-2231571692 when stopping the first message of a chat 🧐

File: backend/danswer/server/query_and_chat/chat_backend.py (hunk 6 above)
Comment (pablonyx) [reply backend/danswer/server/query_and_chat/chat_backend.py]: Might be slightly tricky considering the asynchronous nature of the callback- happy middle ground might be to see the performance impact and then consider doing something like move blocking operations to  a separate thread?

Comment (pablonyx) [reply backend/danswer/server/query_and_chat/chat_backend.py]: Specifically, request.is_disconnected, from the FastAPI docs, returns a coroutine

**Comment by Weves** [comment]
I'm still seeing the weird behavior with stopping the first message:

https://github.com/user-attachments/assets/a47c9ef2-b951-4185-b416-6aeabfd416bf

Also, seems like the message ID is being displayed 🧐

**Comment by Weves** [comment]
Sometimes after stopping a message it disappears 🧐 

https://github.com/user-attachments/assets/5742808d-8774-48a7-9e15-d9c932c44639

Additionally, when you stop the first message, it doesn't give the chat a name. I can see us getting some complaints about this one.

**Review by Weves — APPROVED** [review]
